File: Word Search.cpp
class Solution {
private:
    vector <int> dx {-1, 1, 0, 0}, dy{0, 0, -1, 1};

public:
    bool isValid(int row, int col, vector <vector <char> > &board){
        return row >= 0 and col >= 0 and row < board.size() and col < board[0].size();
    }

    void solve(int pos, int row, int col, vector <vector <int> > &vis, vector <vector <char> > &board, string word, bool &exists){
        if(exists or pos == (int) word.size()){
            exists = true;
            return;
        }
        for(int i = 0; i < 4; ++i){
            int newRow = row + dx[i];
            int newCol = col + dy[i];
            if(isValid(newRow, newCol, board) and vis[newRow][newCol] == false and board[newRow][newCol] == word[pos]){
                vis[newRow][newCol] = true;
                solve(pos + 1, newRow, newCol, vis, board, word, exists);
                vis[newRow][newCol] = false;
            }
        }
    }
    
    bool isWordExist(vector<vector<char>>& board, string word) {
        bool exists = false;
        vector <vector <int> > vis((int) board.size(), vector <int> (board[0].size()));
        for(int i = 0; i < board.size(); ++i){
            for(int j = 0; j < board[0].size(); ++j){
                solve(0, i, j, vis, board, word, exists);
            }
        }
        return exists;
    }
};

File: Word Search.java
class Solution {
    int[] dx = {-1, 0, 1, 0};
    int[] dy = {0, -1, 0, 1};

    boolean validCell(int row, int col, char[][] board){
        return row >= 0 && col >= 0 && row < board.length && col < board[row].length;
    }

    boolean searchWord(int index, int row, int col, char[][] board, String word){
        if(index == word.length()){
            return true;
        }

        if(validCell(row, col, board) == false || word.charAt(index) != board[row][col]){
            return false;
        }

        board[row][col] = '*';
        boolean wordExist = false;
        for(int path = 0; path < 4; ++path){
            wordExist |= searchWord(index + 1, row + dx[path], col + dy[path], board, word);
        }
        board[row][col] = word.charAt(index);

        return wordExist;
    }

    public boolean exist(char[][] board, String word) {
        for(int row = 0; row < board.length; ++row){
            for(int col = 0; col < board[row].length; ++col){
                if(searchWord(0, row, col, board, word)){
                    return true;
                }
            }
        }
        return false;
    }
}
